Feature/reservoirs mvp

westb2 opened the pull request:
Not yet ready for review! Making several final edits and going through commenting pull request for clarity. Will notify when ready.

westb2 commented on docker/dev/Dockerfile:
I would like my docker workflow in the main branch because I believe we should have an image to do dev from, however given that the release image inherits some stuff from here and I want this image to build parflow with the debugger enabled I believe maybe we need a third image. Figured I would open up the disussion.

westb2 commented:
Just realized I will need to figure out how to make reservoirs.names optional or this breaks our test suite

westb2 commented:
Also, should I check in a test case with this for reservoirs?

smithsg84 commented:
Yes, we should have a test case for features.

westb2 commented:
Will be uploading an appropriate test case from my set of personal test cases later, then this should be ready

westb2 commented:
Pending workflow success this PR is ready for review. Since last time, I have added code to support multiple mpi ranks. I have also added two test cases, one for reservoirs as is and one to confirm the work when parallel processing.

westb2 commented:
Just realized at the time of my last comment a push had failed. All changes are in now.

smithsg84 commented on pftools/python/parflow/tools/builders.py:
```diff
@@ -619,9 +633,9 @@ def apply(self, run=None, name_registration=True):
             names_to_set = addon_keys
             for unit_name in valid_unit_names:
-                if unit_name in self.name_registration:
+                if unit_name.casefold() in (registered_name.casefold() for registered_name in self.name_registration):
```
Why is case being removed?   We don't traditionally do that in PF.

smithsg84 commented on test/python/CMakeLists.txt:
```diff
@@ -25,6 +25,8 @@ if(${PARFLOW_HAVE_HYPRE})
     LW_var_dz_spinup
     terrain_following_grid_overland
     var_dz_1D
+    reservoir_mpi_test
```
This doesn't look to be tested in parallel; doesn't look like it is testing under MPI.  Needs to be in the PARALLEL_ topo tests.

smithsg84 commented on test/python/reservoir_mpi_test.py:
```diff
@@ -0,0 +1,345 @@
+
+sloping_slab.FileVersion = 4
+
+sloping_slab.Process.Topology.P = 1
```
Looks to be hard-coding to a single rank run?

smithsg84 commented on pftools/python/parflow/tools/io.py:
```diff
@@ -390,7 +390,7 @@ def __init__(
     ):
         self.filename = file
-        self.f = open(self.filename, 'rb')
+        self.f = open(self.filename, 'rb+')
```
This is a Reader why is writing being added?

smithsg84 commented on pfsimulator/parflow_lib/reservoir_package.c:
```diff
@@ -0,0 +1,533 @@
+              (intake_cell_rank==current_mpi_rank) ||
+              (secondary_intake_cell_rank==current_mpi_rank);
+
+      MPI_Comm new_reservoir_communicator;
```
Will break if MPI is not used.   Need to guard MPI usage.

#ifdef PARFLOW_HAVE_MPI
#else
#endif

The AMPS layer was supposed to hide MPI and do nothing/appropriate when MPI or other parallel layers where not present.   So for things that are not exposed by AMPS we need to guard appropriately or expand the AMPS layer to add the functionality.

smithsg84 commented on pfsimulator/parflow_lib/reservoir_package.c:
```diff
@@ -0,0 +1,533 @@
+ * ReservoirPackage
+ *--------------------------------------------------------------------------*/
+
+bool subgrid_lives_on_this_rank(Subgrid* subgrid, Grid *grid){
```
New code should have doxygen comments for methods.

https://github.com/parflow/parflow/blob/master/CONTRIBUTING.md#source-code-documentation

smithsg84 commented on pfsimulator/parflow_lib/reservoir_package.c:
```diff
@@ -0,0 +1,533 @@
+
+      MPI_Comm new_reservoir_communicator;
+      split_color = part_of_reservoir_lives_on_this_rank ? 1 : MPI_UNDEFINED;
+      MPI_Comm_split(MPI_COMM_WORLD, split_color, current_mpi_rank, &new_reservoir_communicator);
```
Should use amps_CommWorld for MPI_COMM_WORLD.

smithsg84 commented on pfsimulator/parflow_lib/solver_richards.c:
```diff
@@ -1214,7 +1214,10 @@ SetupRichards(PFModule * this_module)
      * Print out the initial well data?
      *-------------------------------------------------------------------*/
-
+    WriteReservoirs("ReservoirsOutput",
```
Add a key to control writing the output, like the print_wells flag.   We allow users to turn off I/O if they don't want it.

smithsg84 commented on pfsimulator/parflow_lib/reservoir.h:
```diff
@@ -0,0 +1,216 @@
+#define RESERVOIRDATA_WRITEHEADER     1
+
+/*     Print Flags      */
+#define RESERVOIRDATA_PRINTPHYSICAL  0x0001
```
Are these flags used somewhere?   I'm not seeing it show up in search.

smithsg84 commented on pfsimulator/parflow_lib/reservoir.h:
```diff
@@ -0,0 +1,216 @@
+    Subgrid       *secondary_intake_subgrid;
+    Subgrid       *release_subgrid;
+    double size;
+    MPI_Comm mpi_communicator;
```
MPI_Comm needs to be guarded, may not have MPI in build.

smithsg84 commented on pfsimulator/parflow_lib/solver_richards.c:
```diff
@@ -3027,7 +3030,130 @@ AdvanceRichards(PFModule * this_module, double start_time,      /* Starting time
+            }
+          }
+        }
+        if (ReservoirDataPhysicalMpiCommunicator(reservoir_data_physical) != MPI_COMM_NULL){
```
This will break if we build without MPI.

smithsg84 reviewed (changes requested):
Some cleanup fixes and there looks like an issue with multiple processors appending to a shared file.

smithsg84 commented:
Test build is failing; tests require the C to be clean.   Should not show any errors when building with 

-DCMAKE_C_FLAGS="-Wall -Werror"

/home/runner/work/parflow/parflow/pfsimulator/parflow_lib/problem_phase_source.c:74:60: error: unused variable ‘reservoir_intake_subgrid’ [-Werror=unused-variable]
   74 |   Subgrid          *subgrid, *well_subgrid, *tmp_subgrid, *reservoir_intake_subgrid, *reservoir_release_subgrid;

westb2 commented:
Thanks for the review Steve. Sorry to make you leave the Doxygen comment on two in a row - I should have gone back and updated this one after your var-dz review. I've addressed most of your comments but am going to dig a little deeper into the safety of printing that file and ensure the build is passing. I will leave another comment once I have done both those things

westb2 commented:
This is ready for review again when you get the time @smithsg84. I think the tests should be passing as well but I don't have permission to run them here. Since you last looked I have made a minor change outside of addressing your comments. Specifically, I added code within parflow to edit the slopes at the reservoir intake cells such that flow would stop. I have functions that I think work for both overland kinematic and overland flow boundary conditions. However, I am actually a bit unclear on what the best way would be to choose which one to do. It seems we might allow people to use both conditions on the same domain. The one place I can find that seems to be doing what I want is in bc_pressure_package.c, where the conditions are set up. Should I do a loop similar to that to check what type of surface each reservoir is on? Thanks in advance.

smithsg84 commented:
@westb2 , PF requires code to compile cleanly to merge.   There are errors showing up in the CI tests because of this.   Using 

 -DCMAKE_C_FLAGS="-Wall -Werror -Wno-unused-result -Wno-unused-function"
 
 Will flag the errors.  This is useful because the warnings do show real errors, like the code in reservoir_package.c :
```
void stop_outlet_flow_at_cell_overland_flow(int i, int j, ProblemData* problem_data, Grid* grid){
  Vector      *slope_x = ProblemDataTSlopeX(problem_data);
  Vector      *slope_y = ProblemDataTSlopeY(problem_data);
  Subvector   *slope_x_subvector ;
  Subvector   *slope_y_subvector;
  int index_slope_x;
  int index_slope_y;
  double *slope_x_ptr = SubvectorData(slope_x_subvector);;
  double *slope_y_ptr = SubvectorData(slope_y_subvector);
```

the subvectors are being used without being initialized.

westb2 commented:
I think those changes must have been fatfingered in my last intermediate commit, or my test cases never would have passed. Regardless, I checked in a fix as well as added a test case for the overland flow boundary condition. Stuff on my end is building, and my tests are passing. So I think this is ready for review, unless somehow tests are failing in workflow

westb2 commented:
As a last update, we now have a key that is required if any reservoirs are in the domain of "Reservoirs.Overland_Flow_Solver" to indicate what solver is being used in the domain. It will throw an error if an unsupported option is provided. Laura and I felt it was the fastest fix that would not require handling weird edge cases or implementing an overcomplicated solution where we looped over the patches

smithsg84 commented:
Compile is still failing.

westb2 commented:
I am not sure why, but I don't see that error on my local build even with that cmake flag. Not sure why, feels like an unused variable should be a system agnostic warning. I will resolve it, but is there a way to trigger the github build on my own? Don't want to keep bugging you for minor stuff like this.

smithsg84 commented:
Are GitHub actions not running if you create a branch?   When I fork the repo and create a branch, pushes to the branch trigger the CI to run.   I thought if you owned the repo you could trigger actions on it.  

https://github.com/westb2/parflow/actions

The merge checks are run as part of the project and the setup was configured to run require a project owner to trigger builds for first time contributors.   Once a contribution is approved, future PR should not require actions.  GitHub setup things this way since people were doing PR's that ran bitcoin like stuff and chewing up compute time.

smithsg84 commented:
Warning and errors can be compiler specific.   The Ubuntu 22.04 build is using GCC GNU 11.4.0.

westb2 commented:
If not it's fine as I need to figure out how to get those warnings showing on my local anyways, but it would be nice to confirm that the build isn't failing for some other reason on remote before asking for more of your time.

westb2 commented:
Hi @smithsg84. Just went through and made sure I had addressed all your changes. There was one comment on a test I wasn't sure what our pattern was (regarding hardcoding number of ranks), but other than that whenever you get a chance to review this again it is ready for you.

smithsg84 commented on pfsimulator/parflow_lib/reservoir_package.c:
```diff
@@ -0,0 +1,594 @@
+      rx = 0;
+      ry = 0;
+      rz = 0;
+      #ifdef PARFLOW_HAVE_MPI
```
General comment , we don't indent preprocessor statements.

smithsg84 commented on pfsimulator/parflow_lib/solver_richards.c:
```diff
@@ -3031,7 +3043,129 @@ AdvanceRichards(PFModule * this_module, double start_time,      /* Starting time
+            GrGeomInLoop(i, j, k, gr_domain, r, ix, iy, iz, nx, ny, nz,
+                         {
+                           ip = SubvectorEltIndex(p_sub_sp, i, j, k);
+                           if (k == (grid_nz - 1))
```
This appears to be an assumption about where top of the domain is?

smithsg84 commented on pfsimulator/parflow_lib/solver_richards.c:
```diff
@@ -3031,7 +3043,129 @@ AdvanceRichards(PFModule * this_module, double start_time,      /* Starting time
+              GrGeomInLoop(i, j, k, gr_domain, r, ix, iy, iz, nx, ny, nz,
+                           {
+                             ip = SubvectorEltIndex(p_sub_sp, i, j, k);
+                             if (k == (grid_nz - 1))
```
Assumption about where top of domain is?

smithsg84 reviewed (changes requested):
Did a more careful review and found some other issues.

Is there an assumption about where the domain top is?   It looks like there is an assumption that the top of domain is at the top of the grid    grid_nx - 1?   Maybe I'm confused.

westb2 commented:
Hi Steve. Thanks for another careful review. I believe I have addressed all of your comments
